osu.Framework/Bindables/BindableList.cs

```diff
@@ -350,14 +338,30 @@ private void replaceRange(int index, int count, IList newItems, BindableList<T>
                 {
                     // Prevent re-adding the item back to the callee.
                     // That would result in a <see cref="StackOverflowException"/>.
-                    if (b != caller)
-                        b.replaceRange(index, count, newItems, this);
+                    b.replaceRange(index, count, newItems, appliedInstances);
                 }
             }
 
             notifyCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Replace, newItems, removedItems, index));
         }
 
+        /// <summary>
+        /// As we are applying non-atomic operations on a potentially complex bindable tree, care needs to be taken to not apply
+        /// the same operation twice to the same bindable instance.
+        ///
+        /// This call tracks all instances which an operation has already been applied to.
+        /// </summary>
+        /// <param name="appliedInstances">A hash set to be passed down the recursive call tree tracking all applied instances.</param>
+        /// <returns>Whether the current instance has already been applied.</returns>
+        private bool checkAlreadyApplied(HashSet<BindableList<T>> appliedInstances)
+        {
+            if (appliedInstances.Contains(this))
+                return true;
+
+            appliedInstances.Add(this);
+            return false;
```

Comment:
Less readable but should be faster.

```suggestion
            return !appliedInstances.Add(this);
```

Reply:
Nice one.